File: advent2017/extras/cxx/src/days/day13.cpp
#include <chrono>
#include <filesystem>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <openssl/sha.h>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

const std::string kExpectedSha = "b59ed1486b6ec731cb7c2f55fdfec971d1157b9411fb823f9ddf0a3839d12cc8";
const std::string kExpectedPart1 = "2604";
const std::string kExpectedPart2 = "3941460";

struct Layer {
    long long depth;
    long long rng;
    long long period;
};

std::string resolve_input(const std::string& provided) {
    if (!provided.empty()) return provided;
    const std::vector<std::string> cands{"advent2017/Day13/d13_input.txt", "Day13/d13_input.txt", "../Day13/d13_input.txt", "../../Day13/d13_input.txt"};
    for (const auto& c : cands) if (std::filesystem::exists(c)) return c;
    throw std::runtime_error("input not found");
}

std::string read_all(const std::string& path) {
    std::ifstream in(path, std::ios::binary);
    if (!in) throw std::runtime_error("failed to read input");
    return std::string((std::istreambuf_iterator<char>(in)), {});
}

std::string sha256_hex(const std::string& data) {
    unsigned char out[SHA256_DIGEST_LENGTH];
    SHA256(reinterpret_cast<const unsigned char*>(data.data()), data.size(), out);
    std::ostringstream oss;
    oss << std::hex << std::setfill('0');
    for (auto b : out) oss << std::setw(2) << static_cast<int>(b);
    return oss.str();
}

std::vector<Layer> parse_layers(const std::string& raw) {
    std::vector<Layer> out;
    std::istringstream in(raw);
    std::string line;
    while (std::getline(in, line)) {
        if (line.empty()) continue;
        const auto pos = line.find(':');
        if (pos == std::string::npos) throw std::runtime_error("bad line");
        const long long d = std::stoll(line.substr(0, pos));
        const long long r = std::stoll(line.substr(pos + 1));
        out.push_back(Layer{d, r, 2 * (r - 1)});
    }
    return out;
}

long long solve_part1(const std::vector<Layer>& layers) {
    long long severity = 0;
    for (const auto& l : layers) {
        if (l.depth % l.period == 0) severity += l.depth * l.rng;
    }
    return severity;
}

long long solve_part2(const std::vector<Layer>& layers) {
    long long delay = 0;
    while (true) {
        bool ok = true;
        for (const auto& l : layers) {
            if ((delay + l.depth) % l.period == 0) {
                ok = false;
                break;
            }
        }
        if (ok) return delay;
        delay++;
    }
}

}  // namespace

int main(int argc, char** argv) {
    int part = 0;
    std::string input;
    for (int i = 1; i < argc; ++i) {
        std::string a = argv[i];
        if (a == "--part") part = std::stoi(argv[++i]);
        else if (a == "--input") input = argv[++i];
        else throw std::runtime_error("unknown arg");
    }
    if (part != 1 && part != 2) throw std::runtime_error("--part 1|2 required");

    input = resolve_input(input);
    const std::string raw = read_all(input);
    if (sha256_hex(raw) != kExpectedSha) throw std::runtime_error("checksum mismatch");
    const auto layers = parse_layers(raw);

    auto t0 = std::chrono::steady_clock::now();
    const std::string ans = (part == 1) ? std::to_string(solve_part1(layers)) : std::to_string(solve_part2(layers));
    const std::string expected = (part == 1) ? kExpectedPart1 : kExpectedPart2;
    if (ans != expected) throw std::runtime_error("answer mismatch");

    std::cout << ans << "\n";
    auto t1 = std::chrono::steady_clock::now();
    const double ms = std::chrono::duration<double, std::milli>(t1 - t0).count();
    std::cerr << "[cxx-fancy] day=13 part=" << part << " runtime_ms=" << ms << "\n";
    return 0;
}

File: advent2017/extras/rust/src/bin/day13.rs
// Architecture Notes:
// - This file is heavily commented for long-term maintainability and reconstruction.
// - Pipeline shape is parse -> model -> compute -> emit.
// - Performance-sensitive sections document data-layout and concurrency tradeoffs.

use sha2::{Digest, Sha256};
use std::env;
use std::fs;
use std::path::Path;
use std::time::Instant;

const EXPECTED_SHA: &str = "b59ed1486b6ec731cb7c2f55fdfec971d1157b9411fb823f9ddf0a3839d12cc8";
const EXPECTED_PART1: &str = "2604";
const EXPECTED_PART2: &str = "3941460";

#[derive(Clone, Copy)]
struct Layer {
    depth: i64,
    rng: i64,
    period: i64,
}

fn resolve_input(provided: &str) -> String {
    if !provided.is_empty() { return provided.to_string(); }
    let cands = ["advent2017/Day13/d13_input.txt", "Day13/d13_input.txt", "../Day13/d13_input.txt", "../../Day13/d13_input.txt"];
    for c in cands { if Path::new(c).exists() { return c.to_string(); } }
    panic!("input not found");
}

fn sha256_file(path: &str) -> String {
    let b = fs::read(path).expect("read");
    let mut h = Sha256::new(); h.update(&b); hex::encode(h.finalize())
}

fn parse_layers(raw: &str) -> Vec<Layer> {
    raw.lines()
        .filter_map(|line| {
            let line = line.trim();
            if line.is_empty() { return None; }
            let parts = line.split(':').map(|x| x.trim()).collect::<Vec<&str>>();
            let d = parts[0].parse::<i64>().expect("int");
            let r = parts[1].parse::<i64>().expect("int");
            Some(Layer { depth: d, rng: r, period: 2 * (r - 1) })
        })
        .collect::<Vec<Layer>>()
}

fn solve_part1(layers: &[Layer]) -> i64 {
    layers
        .iter()
        .filter(|l| l.depth % l.period == 0)
        .map(|l| l.depth * l.rng)
        .sum()
}

fn solve_part2(layers: &[Layer]) -> i64 {
    let mut delay: i64 = 0;
    loop {
        let ok = layers.iter().all(|l| (delay + l.depth) % l.period != 0);
        if ok { return delay; }
        delay += 1;
    }
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let mut part = 0;
    let mut input = String::new();
    let mut i = 1;
    while i < args.len() {
        match args[i].as_str() {
            "--part" => { part = args[i + 1].parse().unwrap(); i += 2; }
            "--input" => { input = args[i + 1].clone(); i += 2; }
            _ => panic!("unknown arg"),
        }
    }
    if part != 1 && part != 2 { panic!("--part 1|2 required"); }

    let in_path = resolve_input(&input);
    if sha256_file(&in_path) != EXPECTED_SHA { panic!("checksum mismatch"); }
    let raw = fs::read_to_string(&in_path).expect("read_to_string");
    let layers = parse_layers(&raw);

    let t0 = Instant::now();
    let ans = if part == 1 { solve_part1(&layers).to_string() } else { solve_part2(&layers).to_string() };
    let expected = if part == 1 { EXPECTED_PART1 } else { EXPECTED_PART2 };
    if ans != expected { panic!("answer mismatch"); }

    println!("{}", ans);
    let ms = t0.elapsed().as_nanos() as f64 / 1e6;
    eprintln!("[rust-fancy] day=13 part={} runtime_ms={:.3}", part, ms);
}
